libsplinter/src/registry/diesel/operations/list_nodes.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides the "list nodes" operation for the `DieselRegistry`.
+
+use diesel::{dsl::sql_query, prelude::*};
+
+use crate::registry::{
+    diesel::{
+        models::{NodeEndpointsModel, NodeKeysModel, NodeMetadataModel, NodesModel},
+        schema::{
+            splinter_nodes, splinter_nodes_endpoints, splinter_nodes_keys, splinter_nodes_metadata,
+        },
+    },
+    MetadataPredicate, Node, NodeBuilder, RegistryError,
+};
+
+use super::{exists_statements_from_metadata_predicates, RegistryOperations};
+
+pub(in crate::registry::diesel) trait RegistryListNodesOperation {
+    fn list_nodes(&self, predicates: &[MetadataPredicate]) -> Result<Vec<Node>, RegistryError>;
```

Comment:
This currently loads all nodes into a `Vec` in memory. Diesel does not currently support iterators of query results. I found a GitHub issue about this (https://github.com/diesel-rs/diesel/issues/1087) with an example implementation of a buffering iterator, but it is not a complete implementation (it doesn't work with simultaneous connections without a lock). This seems like a difficult problem to solve, so I didn't solve it in this PR.